components/calendar/src/astronomy.rs

```diff
@@ -0,0 +1,890 @@
+
+    /// Convert a longitude into a mean time zone;
+    /// this yields the difference in Moment given a longitude
+    /// e.g. a longitude of 90 degrees is 0.25 (90 / 360) days ahead
+    /// of a location with a longitude of 0 degrees.
+    #[allow(dead_code)] // TODO: Remove dead_code tag after use
+    pub(crate) fn zone_from_longitude(longitude: f64) -> f64 {
+        longitude / 360.0
+    }
+
+    /// Convert from local mean time to universal time given a location
+    #[allow(dead_code)] // TODO: Remove dead_code tag after use
+    pub(crate) fn universal_from_local(local_time: Moment, location: Location) -> Moment {
+        local_time - Self::zone_from_longitude(location.longitude)
+    }
+
+    /// Convert from universal time to local time given a location
+    #[allow(dead_code)] // TODO: Remove dead_code tag after use
+    pub(crate) fn local_from_universal(universal_time: Moment, location: Location) -> Moment {
+        universal_time + Self::zone_from_longitude(location.longitude)
+    }
+}
+
+#[derive(Debug)]
+/// The Astronomical struct provides functions which support astronomical
+/// calculations used by many observational calendars.
+pub(crate) struct Astronomical;
+
+impl Astronomical {
+    /// Function for the ephemeris correction, which corrects the
+    /// somewhat-unpredictable discrepancy between dynamical time
+    /// and universal time
+    ///
+    /// Reference code: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L3884-L3952
+    #[allow(dead_code)] // TODO: Remove dead_code tag after use
+    pub(crate) fn ephemeris_correction(moment: Moment) -> f64 {
+        let year = moment.inner() / 365.2425;
+        let year_int = (if year > 0.0 { year + 1.0 } else { year }) as i32;
+        #[allow(clippy::expect_used)]
+        // Month and day are guaranteed to be valid, year is not checked in Iso Date constructor
+        let gregorian: Date<Gregorian> = Date::try_new_gregorian_date(year_int, 7, 1)
+            .expect("Date generation failed for {year} July 1");
+        let iso: Date<Iso> = gregorian.to_iso();
+        let fixed_mid_year: RataDie = Iso::fixed_from_iso(*iso.inner());
+        let c = (1.0 / 36525.0) * ((fixed_mid_year.to_i64_date() as f64) - 693596.0);
+        let y2000 = (year_int - 2000) as f64;
+        let y1700 = (year_int - 1700) as f64;
+        let y1600 = (year_int - 1600) as f64;
+        let y1000 = ((year_int - 1000) as f64) / 100.0;
+        let y0 = year_int as f64 / 100.0;
+        let y1820 = ((year_int - 1820) as f64) / 100.0;
+
+        if (2051..=2150).contains(&year_int) {
+            (8.0 * libm::pow(year_int as f64, 2.0) + 8.0 * libm::pow(1820.0, 2.0)
+                - 30527.0 * (year_int as f64)
+                + 2975050.0)
+                / 216000000.0
+        } else if (2006..=2050).contains(&year_int) {
+            1573.0 / 2160000.0
+                + 10739.0 * y2000 / 2880000000.0
+                + 207.0 * libm::pow(y2000, 2.0) / 3200000000.0
```

Comment:
Was adjusting the constants necessary here? Also, it should give better precision to do the numerator sum/difference and then divide the result at the end. Probably 1 or 2 fewer instructions, FWIW.